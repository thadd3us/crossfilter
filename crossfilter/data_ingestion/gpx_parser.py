"""GPX file parsing functions that convert GPX files to DataFrames."""

import hashlib
import logging
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import List, Optional

import gpxpy
import pandas as pd
from gpxpy.gpx import GPXTrackPoint, GPXWaypoint

from crossfilter.core.schema import DataType, SchemaColumns

logger = logging.getLogger(__name__)


def generate_uuid_from_components(
    first_trackpoint_timestamp: datetime,
    first_trackpoint_lat: float,
    first_trackpoint_lon: float,
    point_timestamp: datetime,
    point_lat: float,
    point_lon: float,
    point_type: DataType,
) -> str:
    """Generate a deterministic UUID from the specified components."""
    # Create a string representation of all components
    components = [
        first_trackpoint_timestamp.isoformat(),
        str(first_trackpoint_lat),
        str(first_trackpoint_lon),
        point_timestamp.isoformat(),
        str(point_lat),
        str(point_lon),
        point_type.value,
    ]
    
    # Create a deterministic hash from the components
    component_string = "|".join(components)
    hash_digest = hashlib.sha256(component_string.encode("utf-8")).digest()
    
    # Create a UUID from the hash
    return str(uuid.UUID(bytes=hash_digest[:16]))


def generate_uuid_for_row(
    row: pd.Series,
    first_point_timestamp: datetime,
    first_point_lat: float,
    first_point_lon: float,
) -> str:
    """Generate a UUID for a single DataFrame row."""
    return generate_uuid_from_components(
        first_point_timestamp,
        first_point_lat,
        first_point_lon,
        row[SchemaColumns.TIMESTAMP_UTC].to_pydatetime().replace(tzinfo=timezone.utc),
        row[SchemaColumns.GPS_LATITUDE],
        row[SchemaColumns.GPS_LONGITUDE],
        DataType(row[SchemaColumns.DATA_TYPE]),
    )


def load_gpx_file_to_df(gpx_file_path: Path) -> pd.DataFrame:
    """
    Parse a GPX file and convert it to a DataFrame conforming to the schema.
    
    Args:
        gpx_file_path: Path to the GPX file
        
    Returns:
        DataFrame with GPX trackpoints and waypoints
        
    Raises:
        FileNotFoundError: If the GPX file doesn't exist
        ValueError: If the GPX file is invalid or empty
    """
    if not gpx_file_path.exists():
        raise FileNotFoundError(f"GPX file not found: {gpx_file_path}")
    
    try:
        with open(gpx_file_path, "r", encoding="utf-8") as f:
            gpx = gpxpy.parse(f)
    except Exception as e:
        raise ValueError(f"Failed to parse GPX file {gpx_file_path}: {e}")
    
    # Get file size
    file_size = gpx_file_path.stat().st_size
    
    # Collect all points first (without UUIDs)
    records: List[dict] = []
    
    # Process all trackpoints
    for track in gpx.tracks:
        track_name = track.name or None
        for segment in track.segments:
            for point in segment.points:
                if point.time and point.latitude and point.longitude:
                    records.append({
                        SchemaColumns.DATA_TYPE: DataType.GPX_TRACKPOINT,
                        SchemaColumns.NAME: track_name,
                        SchemaColumns.CAPTION: None,
                        SchemaColumns.SOURCE_FILE: str(gpx_file_path),
                        SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: point.time.isoformat(),
                        SchemaColumns.TIMESTAMP_UTC: point.time if point.time.tzinfo is not None else point.time.replace(tzinfo=timezone.utc),
                        SchemaColumns.GPS_LATITUDE: point.latitude,
                        SchemaColumns.GPS_LONGITUDE: point.longitude,
                        SchemaColumns.RATING_0_TO_5: None,
                        SchemaColumns.SIZE_IN_BYTES: file_size,
                    })
    
    # Process all waypoints
    for waypoint in gpx.waypoints:
        if waypoint.time and waypoint.latitude and waypoint.longitude:
            records.append({
                SchemaColumns.DATA_TYPE: DataType.GPX_WAYPOINT,
                SchemaColumns.NAME: waypoint.name,
                SchemaColumns.CAPTION: waypoint.comment or waypoint.description,
                SchemaColumns.SOURCE_FILE: str(gpx_file_path),
                SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE: waypoint.time.isoformat(),
                SchemaColumns.TIMESTAMP_UTC: waypoint.time if waypoint.time.tzinfo is not None else waypoint.time.replace(tzinfo=timezone.utc),
                SchemaColumns.GPS_LATITUDE: waypoint.latitude,
                SchemaColumns.GPS_LONGITUDE: waypoint.longitude,
                SchemaColumns.RATING_0_TO_5: None,
                SchemaColumns.SIZE_IN_BYTES: file_size,
            })
    
    if not records:
        logger.warning(f"No valid trackpoints or waypoints found in {gpx_file_path}")
        # Return empty DataFrame with correct schema
        return pd.DataFrame(columns=[
            SchemaColumns.UUID_STRING,
            SchemaColumns.DATA_TYPE,
            SchemaColumns.NAME,
            SchemaColumns.CAPTION,
            SchemaColumns.SOURCE_FILE,
            SchemaColumns.TIMESTAMP_MAYBE_TIMEZONE_AWARE,
            SchemaColumns.TIMESTAMP_UTC,
            SchemaColumns.GPS_LATITUDE,
            SchemaColumns.GPS_LONGITUDE,
            SchemaColumns.RATING_0_TO_5,
            SchemaColumns.SIZE_IN_BYTES,
        ])
    
    # Create DataFrame
    df = pd.DataFrame(records)
    
    # Convert TIMESTAMP_UTC to proper datetime with UTC timezone
    df[SchemaColumns.TIMESTAMP_UTC] = pd.to_datetime(df[SchemaColumns.TIMESTAMP_UTC], utc=True)
    
    # Find the first point for UUID generation (first chronologically)
    first_row = df.loc[df[SchemaColumns.TIMESTAMP_UTC].idxmin()]
    first_point_timestamp = first_row[SchemaColumns.TIMESTAMP_UTC].to_pydatetime().replace(tzinfo=timezone.utc)
    first_point_lat = first_row[SchemaColumns.GPS_LATITUDE]
    first_point_lon = first_row[SchemaColumns.GPS_LONGITUDE]
    
    # Generate UUIDs for all rows using vectorized operation
    df[SchemaColumns.UUID_STRING] = df.apply(
        lambda row: generate_uuid_for_row(row, first_point_timestamp, first_point_lat, first_point_lon),
        axis=1
    )
    
    # Set index name
    df.index.name = SchemaColumns.DF_ID
    
    logger.info(f"Loaded {len(df)} points from {gpx_file_path}")
    
    return df