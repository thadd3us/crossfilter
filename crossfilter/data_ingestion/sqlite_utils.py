"""Shared SQLite utility functions for data ingestion."""

import logging
import re
import uuid
from pathlib import Path
from typing import Any, Union

import pandas as pd
import tqdm
from sqlalchemy import (
    Engine,
    create_engine,
    inspect,
    text,
)

from crossfilter.core.schema import SchemaColumns

logger = logging.getLogger(__name__)


def get_pandas_to_sqlalchemy_dtype(pandas_dtype: str) -> str:
    """Map pandas dtype to SQLAlchemy type string."""
    if pandas_dtype.startswith("int"):
        return "INTEGER"
    elif pandas_dtype.startswith("float"):
        return "REAL"
    elif pandas_dtype.startswith("datetime"):
        return "TIMESTAMP"
    elif pandas_dtype.startswith("bool"):
        return "BOOLEAN"
    else:
        return "TEXT"


def has_unique_constraint_on_uuid(engine: Engine, table_name: str) -> bool:
    """Check if table has a unique constraint or index on UUID_STRING column."""
    inspector = inspect(engine)

    # Check for unique constraints
    try:
        unique_constraints = inspector.get_unique_constraints(table_name)
        for constraint in unique_constraints:
            if SchemaColumns.UUID_STRING in constraint.get("column_names", []):
                return True
    except Exception as e:
        logger.debug(f"get_unique_constraints not supported: {e}")

    # Check for unique indexes
    try:
        indexes = inspector.get_indexes(table_name)
        for index in indexes:
            if index.get("unique", False) and SchemaColumns.UUID_STRING in index.get(
                "column_names", []
            ):
                return True
    except Exception as e:
        logger.debug(f"get_indexes failed: {e}")

    return False


def create_or_update_table_schema(
    engine: Engine, table_name: str, df: pd.DataFrame
) -> None:
    """Create table or add missing columns if table exists."""
    # Validate table name
    if not is_valid_sql_identifier(table_name):
        raise ValueError(f"Invalid table name: {table_name}")

    inspector = inspect(engine)
    with engine.connect() as conn:
        if not inspector.has_table(table_name):
            # Create new empty table
            logger.info(f"Creating new empty table: {table_name}")
            conn.execute(
                text(f"CREATE TABLE {table_name} ({SchemaColumns.UUID_STRING} STRING)")
            )

        # Check if we need to add columns
        existing_columns = {col["name"] for col in inspector.get_columns(table_name)}
        df_columns = set(df.columns)
        missing_columns = df_columns - existing_columns

        if missing_columns:
            logger.info(f"Adding missing columns to {table_name}: {missing_columns}")
            for col in missing_columns:
                if not is_valid_sql_identifier(col):
                    raise ValueError(f"Invalid column name: {col}")
                dtype = get_pandas_to_sqlalchemy_dtype(str(df[col].dtype))
                conn.execute(text(f"ALTER TABLE {table_name} ADD COLUMN {col} {dtype}"))

        # Check if unique constraint exists on UUID_STRING
        has_unique = has_unique_constraint_on_uuid(engine, table_name)
        if not has_unique:
            # Create unique index - let exceptions propagate if there are duplicates
            conn.execute(
                text(
                    f"CREATE UNIQUE INDEX IF NOT EXISTS idx_{table_name}_uuid ON {table_name} ({SchemaColumns.UUID_STRING})"
                )
            )
            logger.info(
                f"Created unique index on {SchemaColumns.UUID_STRING} for table {table_name}"
            )
        conn.commit()
    logger.info(f"Committed table {table_name} with schema {df.columns.tolist()}")


def upsert_dataframe_to_sqlite(
    df: pd.DataFrame, destination_sqlite_db: Path, destination_table: str
) -> None:
    """Upsert DataFrame to SQLite database using bulk operations."""
    if df.columns.empty:
        raise ValueError("DataFrame has no columns, we don't know what schema to use.")

    # Validate table name
    if not is_valid_sql_identifier(destination_table):
        raise ValueError(f"Invalid table name: {destination_table}")

    # Create database directory if it doesn't exist
    destination_sqlite_db.parent.mkdir(parents=True, exist_ok=True)

    # Create engine
    engine = create_engine(f"sqlite:///{destination_sqlite_db}")

    # Create or update table structure - will ensure unique constraint exists
    create_or_update_table_schema(engine, destination_table, df)

    # For bulk upsert, we'll use a temporary table approach
    temp_table = f"{destination_table}_temp_{uuid.uuid4().hex}"

    with engine.connect() as conn:
        # Create temporary table with same structure
        # conn.execute(
        #     text(
        #         f"CREATE TABLE {escape_sql_identifier(temp_table)} AS SELECT * FROM {escape_sql_identifier(destination_table)} WHERE 1=0"
        #     )
        # )

        max_rows_per_insert = 500
        logger.info(
            f"Inserting {len(df)} rows to {temp_table} in batches of {max_rows_per_insert}"
        )
        for i in tqdm.tqdm(range(0, len(df), max_rows_per_insert)):
            df.iloc[i : i + max_rows_per_insert].to_sql(
                temp_table, engine, if_exists="append", index=False, method="multi"
            )
        logger.info(f"Upserted {len(df)} rows to {temp_table}")

        # Perform bulk upsert using a single INSERT OR REPLACE statement
        # This is much more efficient than individual row processing
        columns = list(df.columns)

        # Validate all column names
        for col in columns:
            if not is_valid_sql_identifier(col):
                raise ValueError(f"Invalid column name: {col}")

        columns_str = ", ".join(escape_sql_identifier(col) for col in columns)

        # Single bulk upsert operation
        upsert_stmt = text(
            f"""
            INSERT OR REPLACE INTO {escape_sql_identifier(destination_table)} ({columns_str})
            SELECT {columns_str} FROM {escape_sql_identifier(temp_table)}
        """
        )
        conn.execute(upsert_stmt)

        # Clean up temporary table
        conn.execute(text(f"DROP TABLE {temp_table}"))

        conn.commit()

    logger.info(
        f"Upserted {len(df)} records to {destination_table} in {destination_sqlite_db}"
    )


def query_sqlite_to_dataframe(
    sqlite_db_path: Path, query: str, params: Union[dict[str, Any], None] = None
) -> pd.DataFrame:
    """Execute SQL query against SQLite database and return DataFrame."""
    engine = create_engine(f"sqlite:///{sqlite_db_path}")

    with engine.connect() as conn:
        df = pd.read_sql_query(query, conn, params=params)

    # Workaround: https://github.com/pandas-dev/pandas/issues/55554
    if SchemaColumns.TIMESTAMP_UTC in df.columns:
        df[SchemaColumns.TIMESTAMP_UTC] = pd.to_datetime(
            df[SchemaColumns.TIMESTAMP_UTC], utc=True, format="ISO8601"
        )

    return df


def is_valid_sql_identifier(identifier: str) -> bool:
    """Validate that a string is a safe SQL identifier (column/table name)."""
    # Only allow alphanumeric characters, underscores, and hyphens
    # Must start with a letter or underscore
    pattern = r"^[a-zA-Z_][a-zA-Z0-9_-]*$"
    return bool(re.match(pattern, identifier))


def escape_sql_identifier(identifier: str) -> str:
    """Escape a SQL identifier by wrapping in double quotes."""
    if not is_valid_sql_identifier(identifier):
        raise ValueError(f"Invalid SQL identifier: {identifier}")
    return f'"{identifier}"'
