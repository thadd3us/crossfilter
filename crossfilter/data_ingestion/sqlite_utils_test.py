"""Tests for shared SQLite utility functions."""

import sqlite3
from pathlib import Path
from typing import List

import pandas as pd
import pytest
import sqlalchemy.exc
from sqlalchemy import create_engine, inspect, text
from syrupy.assertion import SnapshotAssertion

from crossfilter.core.schema import DataType, SchemaColumns
from crossfilter.data_ingestion.sqlite_utils import (
    create_or_update_table_schema,
    get_pandas_to_sqlalchemy_dtype,
    has_unique_constraint_on_uuid,
    upsert_dataframe_to_sqlite,
    query_sqlite_to_dataframe,
)


def test_get_pandas_to_sqlalchemy_dtype() -> None:
    """Test mapping pandas dtypes to SQLAlchemy types."""
    assert get_pandas_to_sqlalchemy_dtype("int64") == "INTEGER"
    assert get_pandas_to_sqlalchemy_dtype("float64") == "REAL"
    assert get_pandas_to_sqlalchemy_dtype("datetime64[ns, UTC]") == "TIMESTAMP"
    assert get_pandas_to_sqlalchemy_dtype("bool") == "BOOLEAN"
    assert get_pandas_to_sqlalchemy_dtype("object") == "TEXT"


def test_create_or_update_table_new_table(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test creating a new table."""
    db_path = tmp_path / "test.db"
    engine = create_engine(f"sqlite:///{db_path}")

    # Create test DataFrame
    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749],
            SchemaColumns.GPS_LONGITUDE: [-122.4194],
        }
    )
    create_or_update_table_schema(engine, "test_table", df)
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict() == snapshot


def test_create_or_update_table_add_columns(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test adding columns to existing table."""
    db_path = tmp_path / "test.db"
    engine = create_engine(f"sqlite:///{db_path}")

    initial_df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT],
        }
    )
    create_or_update_table_schema(engine, "test_table", initial_df)
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict() == snapshot

    extended_df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid-2"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749],
            SchemaColumns.GPS_LONGITUDE: [-122.4194],
        }
    )
    create_or_update_table_schema(engine, "test_table", extended_df)
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict() == snapshot


def test_create_or_update_table_nullable_columns(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test handling nullable columns that exist in table but not in DataFrame."""
    db_path = tmp_path / "test.db"
    engine = create_engine(f"sqlite:///{db_path}")

    # Create table with more columns than DataFrame
    with engine.connect() as conn:
        conn.execute(
            text(
                """
            CREATE TABLE test_table (
                UUID_STRING TEXT,
                DATA_TYPE TEXT,
                GPS_LATITUDE REAL,
                GPS_LONGITUDE REAL,
                EXTRA_COLUMN TEXT
            )
        """
            )
        )
        conn.commit()
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict() == snapshot

    # Create DataFrame with fewer columns
    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749],
        }
    )
    create_or_update_table_schema(engine, "test_table", df)
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict() == snapshot


def test_upsert_dataframe_to_sqlite_empty(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test upserting empty DataFrame."""
    db_path = tmp_path / "test.db"
    empty_df = pd.DataFrame()

    upsert_dataframe_to_sqlite(empty_df, db_path, "test_table")
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict() == snapshot


def test_upsert_dataframe_to_sqlite_new_table(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test upserting to new table."""
    db_path = tmp_path / "test.db"

    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid-1", "test-uuid-2"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT, DataType.GPX_WAYPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749, 37.7750],
            SchemaColumns.GPS_LONGITUDE: [-122.4194, -122.4195],
        }
    )
    upsert_dataframe_to_sqlite(df, db_path, "test_table")
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict() == snapshot


def test_upsert_dataframe_to_sqlite_update_existing(
    tmp_path: Path, snapshot: SnapshotAssertion
) -> None:
    """Test upserting to existing table with updates."""
    db_path = tmp_path / "test.db"

    # Insert initial data
    initial_df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid-1", "test-uuid-2"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT, DataType.GPX_WAYPOINT],
            SchemaColumns.GPS_LATITUDE: [1.1, 2.2],
            SchemaColumns.GPS_LONGITUDE: [-10.1, -20.2],
        }
    )
    upsert_dataframe_to_sqlite(initial_df, db_path, "test_table")
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict(orient="records") == snapshot

    # Update with new data (one existing UUID, one new)
    update_df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: [
                "test-uuid-1",
                "test-uuid-3",
            ],  # uuid-1 exists, uuid-3 is new
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT, DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [1.2, 3.3],  # Updated latitude for uuid-1
            SchemaColumns.GPS_LONGITUDE: [-10.2, -30.3],
        }
    )
    upsert_dataframe_to_sqlite(update_df, db_path, "test_table")
    actual = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")
    assert actual.to_dict(orient="records") == snapshot


def test_upsert_dataframe_to_sqlite_with_none_values(tmp_path: Path) -> None:
    """Test upserting DataFrame with None values."""
    db_path = tmp_path / "test.db"

    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid-1"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749],
            SchemaColumns.GPS_LONGITUDE: [-122.4194],
            SchemaColumns.NAME: [None],  # None value
            SchemaColumns.CAPTION: [None],  # None value
        }
    )

    upsert_dataframe_to_sqlite(df, db_path, "test_table")

    # Check data was inserted
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    cursor.execute("SELECT COUNT(*) FROM test_table")
    count = cursor.fetchone()[0]
    conn.close()

    assert count == 1


def test_existing_table_without_unique_constraint(tmp_path: Path) -> None:
    """Test handling of existing table without unique constraint on UUID_STRING."""

    db_path = tmp_path / "test.db"

    # Create a table manually without unique constraint (simulating old database)
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    cursor.execute(
        """
        CREATE TABLE data (
            UUID_STRING TEXT,
            DATA_TYPE TEXT,
            GPS_LATITUDE REAL,
            GPS_LONGITUDE REAL
        )
    """
    )

    # Insert some data with unique UUIDs initially
    cursor.execute(
        "INSERT INTO data VALUES ('uuid-1', 'GPX_TRACKPOINT', 37.7749, -122.4194)"
    )
    cursor.execute(
        "INSERT INTO data VALUES ('uuid-2', 'GPX_TRACKPOINT', 37.7750, -122.4195)"
    )
    conn.commit()
    conn.close()

    # Now try to update the table structure and upsert new data
    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["uuid-3"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7751],
            SchemaColumns.GPS_LONGITUDE: [-122.4196],
        }
    )

    # This should work - the function should handle missing unique constraint
    upsert_dataframe_to_sqlite(df, db_path, "data")

    # Verify the data was inserted
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    cursor.execute("SELECT COUNT(*) FROM data")
    count = cursor.fetchone()[0]
    assert count == 3

    # Verify unique constraint was created
    engine = create_engine(f"sqlite:///{db_path}")
    assert has_unique_constraint_on_uuid(engine, "data")

    conn.close()


def test_existing_table_with_duplicate_uuids(tmp_path: Path) -> None:
    """Test that attempting to create unique constraint on table with duplicates raises an exception."""

    db_path = tmp_path / "test.db"

    # Create a table manually without unique constraint
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    cursor.execute(
        """
        CREATE TABLE data (
            UUID_STRING TEXT,
            DATA_TYPE TEXT,
            GPS_LATITUDE REAL,
            GPS_LONGITUDE REAL
        )
    """
    )

    # Insert duplicate UUIDs
    cursor.execute(
        "INSERT INTO data VALUES ('duplicate-uuid', 'GPX_TRACKPOINT', 37.7749, -122.4194)"
    )
    cursor.execute(
        "INSERT INTO data VALUES ('duplicate-uuid', 'GPX_TRACKPOINT', 37.7750, -122.4195)"
    )
    conn.commit()
    conn.close()

    # Try to upsert new data
    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["new-uuid"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7751],
            SchemaColumns.GPS_LONGITUDE: [-122.4196],
        }
    )

    # This should raise an exception when trying to create unique constraint
    with pytest.raises(sqlalchemy.exc.IntegrityError):
        upsert_dataframe_to_sqlite(df, db_path, "data")


def test_upsert_with_clean_data_works_properly(tmp_path: Path) -> None:
    """Test that upsert works properly when data doesn't have duplicate UUIDs."""

    db_path = tmp_path / "test.db"

    # First, insert some clean data
    df1 = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["uuid-1", "uuid-2"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT, DataType.GPX_TRACKPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749, 37.7750],
            SchemaColumns.GPS_LONGITUDE: [-122.4194, -122.4195],
        }
    )

    upsert_dataframe_to_sqlite(df1, db_path, "data")

    # Now try to upsert data that updates existing record and adds new one
    df2 = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: [
                "uuid-1",
                "uuid-3",
            ],  # Update uuid-1, add uuid-3
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT, DataType.GPX_WAYPOINT],
            SchemaColumns.GPS_LATITUDE: [
                37.7799,
                37.7800,
            ],  # Updated latitude for uuid-1
            SchemaColumns.GPS_LONGITUDE: [-122.4200, -122.4201],
        }
    )

    upsert_dataframe_to_sqlite(df2, db_path, "data")

    # Verify the data
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()

    # Should have 3 total records (uuid-1 updated, uuid-2 unchanged, uuid-3 new)
    cursor.execute("SELECT COUNT(*) FROM data")
    count = cursor.fetchone()[0]
    assert count == 3

    # Check that uuid-1 was updated
    cursor.execute("SELECT GPS_LATITUDE FROM data WHERE UUID_STRING = 'uuid-1'")
    updated_lat = cursor.fetchone()[0]
    assert updated_lat == 37.7799

    # Check that uuid-2 is unchanged
    cursor.execute("SELECT GPS_LATITUDE FROM data WHERE UUID_STRING = 'uuid-2'")
    unchanged_lat = cursor.fetchone()[0]
    assert unchanged_lat == 37.7750

    # Check that uuid-3 was inserted
    cursor.execute("SELECT COUNT(*) FROM data WHERE UUID_STRING = 'uuid-3'")
    new_count = cursor.fetchone()[0]
    assert new_count == 1

    conn.close()


def test_query_sqlite_to_dataframe(tmp_path: Path) -> None:
    """Test querying SQLite database to DataFrame."""
    db_path = tmp_path / "test.db"

    # Create test data
    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid-1", "test-uuid-2"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT, DataType.GPX_WAYPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749, 37.7750],
            SchemaColumns.GPS_LONGITUDE: [-122.4194, -122.4195],
        }
    )

    upsert_dataframe_to_sqlite(df, db_path, "test_table")

    # Query the data back
    result_df = query_sqlite_to_dataframe(db_path, "SELECT * FROM test_table")

    assert len(result_df) == 2
    assert SchemaColumns.UUID_STRING in result_df.columns
    assert SchemaColumns.DATA_TYPE in result_df.columns
    assert SchemaColumns.GPS_LATITUDE in result_df.columns
    assert SchemaColumns.GPS_LONGITUDE in result_df.columns


def test_query_sqlite_to_dataframe_with_params(tmp_path: Path) -> None:
    """Test querying SQLite database with parameters."""
    db_path = tmp_path / "test.db"

    # Create test data
    df = pd.DataFrame(
        {
            SchemaColumns.UUID_STRING: ["test-uuid-1", "test-uuid-2"],
            SchemaColumns.DATA_TYPE: [DataType.GPX_TRACKPOINT, DataType.GPX_WAYPOINT],
            SchemaColumns.GPS_LATITUDE: [37.7749, 37.7750],
            SchemaColumns.GPS_LONGITUDE: [-122.4194, -122.4195],
        }
    )

    upsert_dataframe_to_sqlite(df, db_path, "test_table")

    # Query with parameters
    result_df = query_sqlite_to_dataframe(
        db_path,
        f"SELECT * FROM test_table WHERE {SchemaColumns.UUID_STRING} = :uuid",
        {"uuid": "test-uuid-1"},
    )

    assert len(result_df) == 1
    assert result_df.iloc[0][SchemaColumns.UUID_STRING] == "test-uuid-1"
