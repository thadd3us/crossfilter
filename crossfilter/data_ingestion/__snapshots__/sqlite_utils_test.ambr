# serializer version: 1
# name: test_create_or_update_table_add_columns
  dict({
    'DATA_TYPE': dict({
    }),
    'UUID_STRING': dict({
    }),
  })
# ---
# name: test_create_or_update_table_add_columns.1
  dict({
    'DATA_TYPE': dict({
    }),
    'GPS_LATITUDE': dict({
    }),
    'GPS_LONGITUDE': dict({
    }),
    'UUID_STRING': dict({
    }),
  })
# ---
# name: test_create_or_update_table_new_table
  dict({
    'DATA_TYPE': dict({
    }),
    'GPS_LATITUDE': dict({
    }),
    'GPS_LONGITUDE': dict({
    }),
    'UUID_STRING': dict({
    }),
  })
# ---
# name: test_create_or_update_table_nullable_columns
  dict({
    'DATA_TYPE': dict({
    }),
    'EXTRA_COLUMN': dict({
    }),
    'GPS_LATITUDE': dict({
    }),
    'GPS_LONGITUDE': dict({
    }),
    'UUID_STRING': dict({
    }),
  })
# ---
# name: test_create_or_update_table_nullable_columns.1
  dict({
    'DATA_TYPE': dict({
    }),
    'EXTRA_COLUMN': dict({
    }),
    'GPS_LATITUDE': dict({
    }),
    'GPS_LONGITUDE': dict({
    }),
    'UUID_STRING': dict({
    }),
  })
# ---
# name: test_upsert_dataframe_to_sqlite_empty
  dict({
    'UUID_STRING': dict({
    }),
  })
# ---
# name: test_upsert_dataframe_to_sqlite_new_table
  dict({
    'DATA_TYPE': dict({
      0: 'GPX_TRACKPOINT',
      1: 'GPX_WAYPOINT',
    }),
    'GPS_LATITUDE': dict({
      0: 37.7749,
      1: 37.775,
    }),
    'GPS_LONGITUDE': dict({
      0: -122.4194,
      1: -122.4195,
    }),
    'UUID_STRING': dict({
      0: 'test-uuid-1',
      1: 'test-uuid-2',
    }),
  })
# ---
# name: test_upsert_dataframe_to_sqlite_update_existing
  list([
    dict({
      'DATA_TYPE': 'GPX_TRACKPOINT',
      'GPS_LATITUDE': 1.1,
      'GPS_LONGITUDE': -10.1,
      'UUID_STRING': 'test-uuid-1',
    }),
    dict({
      'DATA_TYPE': 'GPX_WAYPOINT',
      'GPS_LATITUDE': 2.2,
      'GPS_LONGITUDE': -20.2,
      'UUID_STRING': 'test-uuid-2',
    }),
  ])
# ---
# name: test_upsert_dataframe_to_sqlite_update_existing.1
  list([
    dict({
      'DATA_TYPE': 'GPX_WAYPOINT',
      'GPS_LATITUDE': 2.2,
      'GPS_LONGITUDE': -20.2,
      'UUID_STRING': 'test-uuid-2',
    }),
    dict({
      'DATA_TYPE': 'GPX_TRACKPOINT',
      'GPS_LATITUDE': 1.2,
      'GPS_LONGITUDE': -10.2,
      'UUID_STRING': 'test-uuid-1',
    }),
    dict({
      'DATA_TYPE': 'GPX_TRACKPOINT',
      'GPS_LATITUDE': 3.3,
      'GPS_LONGITUDE': -30.3,
      'UUID_STRING': 'test-uuid-3',
    }),
  ])
# ---
